File: Utilities/DatabaseUpdater/ObjectScript.vb
Public Class ObjectScript
    Inherits Script





	Const ExtendedPropertyForLastModifiedName As String = "LastModified"

	Sub New(ByVal path As String, ByVal objectType As String)
		MyBase.New(path)

		Dim filename As String = GetFileName()

		Me._objectName = filename.Substring(0, filename.LastIndexOf("."))

		Me.objectType = objectType
	End Sub
	Public Overrides Function ShouldBeApplied() As Boolean
		Return IsDifferentToTheOneInTheDatabase()
	End Function
	Private Function IsDifferentToTheOneInTheDatabase() As Boolean
		Try
			Dim dbDate As DateTime = DateTime.Parse(SQLExtendedPropertyMethods.Get(ExtendedPropertyForLastModifiedName, Me._objectName, Me.objectType))
			Return Math.Abs(modified.Subtract(CType(dbDate, DateTime)).TotalSeconds) > 1
		Catch ex As NoDatabaseObjectWithThatNameException
			Return True
		End Try
	End Function
	Public Overrides Sub LogUpdateInDatabase()
		SQLExtendedPropertyMethods.Update(ExtendedPropertyForLastModifiedName, Me.modified.ToString, Me._objectName, Me.objectType)
	End Sub

	ReadOnly Property ObjectName() As String
		Get
			Return Me._objectName
		End Get
	End Property

End Class

File: Utilities/SqlScriptRunner/ScriptTypes/ObjectScript.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Common.Automation.Sql;

namespace SqlScriptRunner.ScriptTypes
{
    internal  class ObjectScript :  Script
    {
        const string ExtendedPropertyForLastModifiedName = "LastModified";

        internal ObjectScript(string path, string scriptType)
            : base(path, scriptType)
        {
            this._objectName = this.File.Name.Substring(0, this.File.Name.LastIndexOf("."));
        }

        internal  override bool ShouldBeApplied()
        {
            return IsDifferentToTheOneInTheDatabase();
        }

        private bool IsDifferentToTheOneInTheDatabase()
        {
            try
            {
				DateTime dbDate = DateTime.Parse(SqlExtendedPropertyMethods.GetObjectLevelExtendedProperty(Global.GetConnection(), ExtendedPropertyForLastModifiedName, this._objectName, this.ScriptType));
                return Math.Abs(this.File.LastWriteTime.Subtract(((DateTime)(dbDate))).TotalSeconds) > 1;
            }
            catch (SqlExtendedPropertyMethods.ObjectDoesNotExist)
            {
                return true;
            }
        }

        internal  override void LogUpdateInDatabase()
        {
			SqlExtendedPropertyMethods.Update(Global.GetConnection(), ExtendedPropertyForLastModifiedName, this.File.LastWriteTime.ToString(), this._objectName, this.ScriptType);
        }

        internal string ObjectName
        {
            get
            {
                return this._objectName;
            }
        }
    }
}
